Avoid compiling rules every time the extension is loaded if the rules haven't changed

b-weinstein opened the pull request:
#### 10ce6abfd3467908a62bf2d8ab1e57ac4b356a28
<pre>
Avoid compiling rules every time the extension is loaded if the rules haven&apos;t changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=265473">https://bugs.webkit.org/show_bug.cgi?id=265473</a>
<a href="https://rdar.apple.com/118839289">rdar://118839289</a>

Reviewed by Timothy Hatcher.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::declarativeNetRequestRuleStore):
(WebKit::computeStringHashForContentBlockerRules): Compute the hash for these content blocker rules, making sure to
append the current version of the rule translator, so when it is updated, rules will be re-compiled.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Before compiling the rule, perform a lookup and check
the hash.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: <a href="https://commits.webkit.org/271248@main">https://commits.webkit.org/271248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf1a6062d7fea426095b7f204da2b110814c9850

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29056 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30027 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3839 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28075 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5194 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/23852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/4510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30666 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25398 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/25297 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6204 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6671 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2697,15 +2712,33 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    auto *hashOfWebKitRules = computeStringHashForContentBlockerRules(webKitRules);
+
+    [declarativeNetRequestRuleStore() lookUpContentRuleListForIdentifier:uniqueIdentifier() completionHandler:^(WKContentRuleList *foundRuleList, NSError *) {
+        if (foundRuleList) {
```
should we be logging an error if the lookup failed?

b-weinstein replied:
That can happen for normal reasons, like if it's the first time an extension was loaded, or if the user deleted the file on disk.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -78,10 +78,14 @@
 static constexpr NSInteger currentBackgroundContentListenerStateVersion = 2;
 
+// Update this value when any changes are made to the rule translation logic in _WKWebExtensionDeclarativeNetRequestRule.
+static NSString * const currentDeclarativeNetRequestRuleTranslatorVersion = @"1";
```
I think a `size_t` or other integer type for this would be cleaner, since we are already using a format string to append it.

xeenon replied:
Or `NSInteger` to match the one above.

b-weinstein committed:
Avoid compiling rules every time the extension is loaded if the rules haven't changed
https://bugs.webkit.org/show_bug.cgi?id=265473
rdar://118839289

Reviewed by Timothy Hatcher.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::declarativeNetRequestRuleStore):
(WebKit::computeStringHashForContentBlockerRules): Compute the hash for these content blocker rules, making sure to
append the current version of the rule translator, so when it is updated, rules will be re-compiled.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Before compiling the rule, perform a lookup and check
the hash.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionDeclarativeNetRequestRule.mm:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: https://commits.webkit.org/271248@main
